src/appleseed.studio/mainwindow/rendering/renderwidget.cpp

```diff
@@ -422,5 +428,29 @@ void RenderWidget::paintEvent(QPaintEvent* event)
     m_painter.end();
 }
 
+void RenderWidget::dragEnterEvent(QDragEnterEvent* event)
+{
+    if (event->mimeData()->hasFormat("text/plain") && m_project.has_trace_context())
+        event->acceptProposedAction();
+}
+
+void RenderWidget::dragMoveEvent(QDragMoveEvent* event)
+{
+    if (event->pos().x() > pos().x() + width() || event->pos().y() > pos().y() + height())
+        event->ignore();
+    else
+        event->accept();
+}
+
+void RenderWidget::dropEvent(QDropEvent* event)
+{
+    const Vector2d drop_pos = Vector2d(
```

Comment:
You actually don't need the `Vector2()` part:

    const Vector2d drop_pos(
        static_cast<double>(event->pos().x()) / width(),
        static_cast<double>(event->pos().y()) / height());

Reply:
I dont understand what you mean, there is no implicit conversion between QPoint and vector2d

Reply:
What I mean is that you don't need to create a temporary and assign it to your variable, you can simply construct the `Vector2d` directly.